src/Icicle/Sea/FromAvalanche/Analysis.hs

```diff
@@ -168,3 +173,61 @@ outputsOfStatement stmt
 
      Output n t xts
       -> Map.singleton n (t, fmap snd xts)
+
+------------------------------------------------------------------------
+
+typesOfProgram :: Program (Annot a) n Prim -> Set ValType
+typesOfProgram = typesOfStatement . statements
+
+typesOfStatement :: Statement (Annot a) n Prim -> Set ValType
+typesOfStatement stmt
+ = case stmt of
+     Block []              -> Set.empty
+     Block (s:ss)          -> typesOfStatement s  `Set.union`
+                              typesOfStatement (Block ss)
+     Let _ x ss            -> typesOfExp       x  `Set.union`
+                              typesOfStatement ss
+     If x tt ee            -> typesOfExp       x  `Set.union`
+                              typesOfStatement tt `Set.union`
+                              typesOfStatement ee
+     ForeachInts  _ f t ss -> typesOfExp       f `Set.union`
+                              typesOfExp       t `Set.union`
+                              typesOfStatement ss
+     ForeachFacts _ _ _ ss -> typesOfStatement ss
+     InitAccumulator (Accumulator _ at x) ss
+                           -> Set.singleton    at `Set.union`
+                              typesOfExp       x  `Set.union`
+                              typesOfStatement ss
+
+     Write _ x             -> typesOfExp x
+     LoadResumable _ vt    -> Set.singleton vt
+     SaveResumable _ vt    -> Set.singleton vt
+     Output _ vt xs        -> Set.singleton vt `Set.union`
+                              Set.unions (fmap goOut xs)
+     KeepFactInHistory     -> Set.empty
+
+     Read _ _ vt ss
+      -> Set.singleton vt `Set.union`
+         typesOfStatement ss
```

Comment:
I hope this is just as fast, I use this pattern all the time, containers yields such beautiful code in this arrangement :)